### ISSUE TITLE
Inconsistent getters and setters

### ISSUE THREAD

**Issue by kbacha** [elastic/elasticsearch-dsl-ruby#7]
Throughout the DSL there are many different ways to setting data on a search.

Sometimes it's through a setter:

```ruby
search = Search.new
search.filter = Filter.new
```

Sometimes that setter doesn't exist and setting is based on arguments ONLY
```ruby
search = Search.new
search.sort   # => nil # Only returns the value since no arguments
search.sort {} # Empty block creates the sort object
search.sort # => Sort
```

Sometimes the setter returns some random internal value (if it's a 
```ruby
search = Search.new
search.filter = Filter.new
#
# This will append the filter but then returns the array inside the filter
search.filter.range(:name) # => [Range()]
```

And yet other times it's even more confusing:
https://github.com/elastic/elasticsearch-ruby/blob/master/elasticsearch-dsl/lib/elasticsearch/dsl/search/base_component.rb#L56

Basically, the library becomes very hard to use as either a DSL or as an imperative library.

It's hard to use it as a DSL because it changes the implicit receiver making using it inside another class very hard because almost everything causes method_missing errors when trying to call my own functions.

Using it imperatively is very hard and requires the source code handy because every method call has a totally different, undocumented, argument list and/or a different return value.
